presto-accumulo/src/main/java/com/facebook/presto/accumulo/AccumuloClient.java

```diff
@@ -0,0 +1,1137 @@
+
+                    if (matched.size() != 1) {
+                        throw new PrestoException(StandardErrorCode.USER_ERROR, "Unknown column in locality group: " + col);
+                    }
+
+                    if (matched.get(0).getName().toLowerCase().equals(rowIdColumn)) {
+                        throw new PrestoException(StandardErrorCode.USER_ERROR, "Row ID column cannot be in a locality group");
+                    }
+                }
+            }
+        }
+
+        // Get the column mappings
+        Map<String, Pair<String, String>> mapping = AccumuloTableProperties.getColumnMapping(meta.getProperties());
+        if (mapping == null) {
+            if (AccumuloTableProperties.isExternal(tableProperties)) {
+                throw new PrestoException(StandardErrorCode.USER_ERROR, "Column generation for external tables is not supported, must specify " + AccumuloTableProperties.COLUMN_MAPPING);
+            }
+
+            mapping = autoGenerateMapping(meta.getColumns(), groups);
+        }
+
+        // The list of indexed columns
+        List<String> indexedColumns = AccumuloTableProperties.getIndexColumns(tableProperties);
+
+        // And now we parse the configured columns and create handles for the
+        // metadata manager
+        List<AccumuloColumnHandle> columns = new ArrayList<>();
+        for (int ordinal = 0; ordinal < meta.getColumns().size(); ++ordinal) {
+            ColumnMetadata cm = meta.getColumns().get(ordinal);
+
+            // Special case if this column is the
+            if (cm.getName().toLowerCase().equals(rowIdColumn)) {
+                columns.add(new AccumuloColumnHandle("accumulo", rowIdColumn, null, null,
+                        cm.getType(), ordinal, "Accumulo row ID", false));
+            }
+            else {
+                if (!mapping.containsKey(cm.getName())) {
+                    throw new InvalidParameterException(String
+                            .format("Misconfigured mapping for presto column %s", cm.getName()));
+                }
+
+                // Get the mapping for this column
+                Pair<String, String> famqual = mapping.get(cm.getName());
+                boolean indexed = indexedColumns.contains(cm.getName().toLowerCase());
+                String comment = String.format("Accumulo column %s:%s. Indexed: %b",
+                        famqual.getLeft(), famqual.getRight(), indexed);
+
+                // Create a new AccumuloColumnHandle object
+                columns.add(new AccumuloColumnHandle("accumulo", cm.getName(), famqual.getLeft(),
+                        famqual.getRight(), cm.getType(), ordinal, comment, indexed));
+            }
+        }
+
+        // Finally create the AccumuloTable object
+        boolean external = AccumuloTableProperties.isExternal(tableProperties);
+        AccumuloTable table = new AccumuloTable(meta.getTable().getSchemaName(),
```

Comment:
Wouldn't it be useful to introduce a separate type for external table? E.g. `AccumuloExternalTable` which derives from `AccumuloTable` (base).
It would let you remove this `external` boolean flag in c-tor + would make below flow much simpler, without recurrences `(if (external) { some_checks } register(table); if (external) done; else ... )`
`MetadataManager` would refuse registering unhealthy table (below check would be moved to `MetadataManager.createTableMetadata(AccumuloExternalTable)`). You have `Connector` in `ZookeeperMetadataManager` already (but maybe you don't want to have it there [AD https://github.com/prestodb/presto/pull/5030/files#r59881666]), so this change should not be painful.